File: Java/src/main/java/QuestionBank/Solution001_100/Solution34.java
package QuestionBank.Solution001_100;

import java.util.Arrays;

/**
 * @author windows
 */
public class Solution34 {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0){
            return new int[]{-1, -1};
        }
        int left = -1;
        int right = nums.length;
        while(left != right - 1){
            int mid = left + (right - left) / 2;
            if(nums[mid] < target){
                left = mid;
            }else{
                right = mid;
            }
        }
        if (nums[right] != target){
            return new int[]{-1, -1};
        }
        int[] res = new int[2];
        res[0] = right;
        left = -1;
        right = nums.length;
        while(left != right - 1){
            int mid = left + (right - left) / 2;
            if(nums[mid] <= target){
                left = mid;
            }else{
                right = mid;
            }
        }
        res[1] = left;
        return res;
    }


    public static void main(String[] args) {
        Solution34 solution34 = new Solution34();
        System.out.println(Arrays.toString(solution34.searchRange(new int[]{2, 2}, 3)));
    }
}

File: C++/src/Solution1_100/Solution34.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution34.h"
#include "vector"
using namespace std;

vector<int> searchRange(vector<int>& nums, int target) {
  int  left = -1, right = nums.size();
  while (left < right - 1){
    int mid = left + (right - left) / 2;
    if (nums[mid] < target){
      left = mid;
    }else {
      right = mid;
    }
  }
  int a = left;
  if (a + 1 == nums.size() || nums[a + 1] != target){
    return {-1, -1};
  }
  left = -1, right = nums.size();
  while (left < right - 1){
    int mid = left + (right - left) / 2;
    if (nums[mid] > target){
      right = mid;
    }else {
      right = mid;
    }
  }
  int b = right;
  return {a, b};
}
